biosys/apps/main/utils_data_package.py

```diff
@@ -1081,6 +997,226 @@ def _find_site_code_field(self):
+                self.errors.append(msg)
+        else:
+            # Genus => Multi-column name
+            # We must have a 'Species' column
+            if self.species_field is None:
+                msg = "A field '{genus_col_name}' exists without a {species_col_name}. If the species field is named " \
+                      "differently, you can 'tag' it with the biosys type: '{species_biosys_type}'" \
+                    .format(genus_col_name=SpeciesObservationSchema.GENUS_FIELD_NAME,
+                            species_col_name=SpeciesObservationSchema.SPECIES_NAME_FIELD_NAME,
+                            species_biosys_type=BiosysSchema.SPECIES_TYPE_NAME
+                            )
+                self.errors.append(msg)
+
+        # check required constraint.
+        self.required_errors = self._check_required_constraint()
+
+        self.errors += self.required_errors
+
+    def is_valid(self):
+        return not bool(self.errors)
+
+    @property
+    def valid(self):
+        return self.is_valid()
+
+    @property
+    def has_species_name(self):
+        return self.species_name_field is not None
+
+    @property
+    def is_species_name_only(self):
+        return all([
+            self.has_species_name,
+            not self.has_genus_and_species,
+            not self.has_name_id
+        ])
+
+    @property
+    def has_genus_and_species(self):
+        return self.genus_field is not None and self.species_field is not None
+
+    @property
+    def is_genus_and_species_only(self):
+        return all([
+            self.has_genus_and_species,
+            not self.has_species_name,
+            not self.has_name_id
+        ])
+
+    @property
+    def is_name_id_only(self):
+        return all([
+            self.has_name_id,
+            not self.has_species_name,
+            not self.has_genus_and_species
+        ])
+
+    @property
+    def has_name_id(self):
+        return self.name_id_field
+
+    def cast_species_name_id(self, record):
+        field = self.name_id_field
+        return field.cast(record.get(field.name)) if field is not None else None
+
+    def cast_species_name(self, record):
+        composite = self._compose_species_name(record) if self.has_genus_and_species else None
+        species_name = self._cast_field(self.species_name_field, record) if self.species_name_field else None
+        return composite or species_name
+
+    def _compose_species_name(self, record):
+        """
+        genus + " " + species + " " + infra_rank + " " + infra_name
+        see https://decbugs.com/view.php?id=6674
+        :return: Will raise an exception if constraints are not respected
+        """
+        result = ''
+        for field in [self.genus_field, self.species_field, self.infra_rank_field, self.infra_name_field]:
+            value = self._cast_field(field, record)
+            if value:
+                result += ' {}'.format(value)
+        return result.strip()
```

Comment:
Could alternatively string ```join()``` function here rather than appending strings then stripping.

Reply:
Yes the `join`. I forgot. I was looking for a `reduce` function